### PR TITLE
feat: Google Calendar View/Create Tools

### PR DESCRIPTION
**Pull request by ShaunBaker**
This pull request provides two separate tools that provide the ability for users to interact with a Google Calendar to retrieve and create events in the calendar through input.

### CODE REVIEW

File: langchain/src/tools/google_calendar/base.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      );
+    }
+
+    this.clientEmail = fields.credentials.clientEmail;
+    this.privateKey = fields.credentials.privateKey;
+    this.calendarId = fields.credentials.calendarId;
+    this.scopes = fields.scopes || [];
+  }
+
+  getModel() {
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/base.ts:71]: This should be passed in

File: langchain/src/tools/google_calendar/create.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  constructor(fields: GoogleCalendarAgentParams) {
+    super(fields);
+  }
+
+  async _call(query: string) {
+    const auth = await this.getAuth();
+    const model = this.getModel();
+
+    return await runCreateEvent(query, {
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/create.ts:18]: `await` is unnecessary for return statements

File: langchain/src/tools/google_calendar/descriptions.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+export const CREATE_TOOL_DESCRIPTION = `A tool for creating Google Calendar events and meetings.
+
+INPUT example:
+"action": "google_calendar_create",
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/descriptions.ts:4]: Output parser is the standard way of doing this:

https://js.langchain.com/docs/modules/prompts/output_parsers/#structured-output-parser

File: langchain/src/tools/google_calendar/prompts/create-event-prompt.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+Today's datetime on UTC time 2023-05-02T10:00:00+00:00, it's Tuesday and timezone
+of the user is -5, take into account the timezone of the user and today's date.
+1. event_summary 
+2. event_start_time 
+3. event_end_time 
+4. event_location 
+5. event_description 
+6. user_timezone
+event_summary:
+{{
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/prompts/create-event-prompt.ts:14]: See above

File: langchain/src/tools/google_calendar/prompts/view-events-prompt.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+Today's datetime on UTC time 2023-05-02T10:00:00+00:00, it's Tuesday and timezone
+of the user is -5, take into account the timezone of the user and today's date.
+If the user is searching for events with a specific title, person or location, put it into the search_query parameter.
+1. time_min 
+2. time_max
+3. user_timezone
+4. max_results 
+5. search_query 
+event_summary:
+{{
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/prompts/view-events-prompt.ts:14]: See above

File: langchain/package.json [hunk 6]
```diff
@@ -549,6 +549,7 @@
     "epub2": "^3.0.1",
     "faiss-node": "^0.2.1",
     "google-auth-library": "^8.8.0",
+    "googleapis": "^118.0.0",
```
Comment (jacoblee93) [inline langchain/package.json:552]: Add this to `devDependencies` above as well

File: langchain/src/tools/google_calendar/commands/run-create-events.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
+import { google, calendar_v3 } from "googleapis";
+import type { JWT, GaxiosResponse } from "googleapis-common";
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/commands/run-create-events.ts:2]: This should be an explicit dep?

**Review by jacoblee93 — CHANGES_REQUESTED** [review]
Looks cool! A few questions

File: langchain/src/tools/google_calendar/prompts/view-events-prompt.ts (hunk 5 above)
Comment (ShaunBaker) [reply langchain/src/tools/google_calendar/prompts/view-events-prompt.ts:14]: https://github.com/nearform/langchainjs/blob/main/langchain/src/tools/google_calendar/commands/run-view-events.ts#L21

File: langchain/src/tools/google_calendar/prompts/create-event-prompt.ts (hunk 4 above)
Comment (ShaunBaker) [reply langchain/src/tools/google_calendar/prompts/create-event-prompt.ts:14]: This is passed here: https://github.com/nearform/langchainjs/blob/main/langchain/src/tools/google_calendar/commands/run-create-events.ts#L71

File: langchain/src/tools/google_calendar/base.ts (hunk 1 above)
Comment (jacoblee93) [reply langchain/src/tools/google_calendar/base.ts:77]: This should accept a `BaseLLM` instance - what if someone needed to use a custom model?

File: langchain/src/tools/google_calendar/view.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+  constructor(fields: GoogleCalendarAgentParams) {
+    super(fields);
+  }
+
+  async _call(query: string) {
+    const auth = await this.getAuth();
+    const model = this.getModel();
+
+    return await runViewEvents(query, {
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/view.ts:19]: Don't need to `return await`

File: langchain/src/tools/google_calendar/descriptions.ts (hunk 3 above)
Comment (jacoblee93) [reply langchain/src/tools/google_calendar/descriptions.ts:4]: Ah ok, my mistake!

File: langchain/src/tools/google_calendar/prompts/create-event-prompt.ts (hunk 4 above)
Comment (jacoblee93) [reply langchain/src/tools/google_calendar/prompts/create-event-prompt.ts:14]: In this case using an output parser would make sense I think?

**Review by jacoblee93 — CHANGES_REQUESTED** [review]
Getting there! A few small questions/comments.

**Comment by simoneb** [comment]
@jacoblee93 with the exception of the conflicting files, the feedback should be addressed if you wouldn't mind taking another look please.

File: langchain/src/tools/google_calendar/base.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      credentials: {
+        clientEmail: getEnvironmentVariable("GOOGLE_CALENDAR_CLIENT_EMAIL"),
+        privateKey: getEnvironmentVariable("GOOGLE_CALENDAR_PRIVATE_KEY"),
+        calendarId: getEnvironmentVariable("GOOGLE_CALENDAR_CALENDAR_ID"),
+      },
+      scopes: [
+        "https://www.googleapis.com/auth/calendar",
+        "https://www.googleapis.com/auth/calendar.events",
+      ],
+      openAIParameters: {
```
Comment (jacoblee93) [inline langchain/src/tools/google_calendar/base.ts:44]: Your input should not take `openAIParameters`, but an actual BaseLLM instance that you then set as `this.llm`. Someone could use this with e.g. GPT4All or Anthropic. 

Also, I don't think you're actually initializing a model instance anyway? Does this pass your test cases?

**Review by jacoblee93 — CHANGES_REQUESTED** [review]
Can you test this end-to-end as well please?

**Pull request closed without merging**